hicexplorer/hicBuildMatrix.py

```diff
@@ -822,74 +844,363 @@ def main(args=None):
+                                count_left, count_right, inter_chromosomal, short_range, long_range, pair_added, len(pMateBuffer1), pResultIndex, pCounter]])
+    return
+
+
+def write_bam(pOutputBam, pTemplate, pOutputFileBufferDir):
+
+    out_bam = pysam.Samfile(pOutputFileBufferDir + pOutputBam, 'wb', template=pTemplate)
+
+    counter = 0
+    while not os.path.isfile(pOutputFileBufferDir + 'done_processing') or os.path.isfile(pOutputFileBufferDir + str(counter) + '.bam_done'):
+        if os.path.isfile(pOutputFileBufferDir + str(counter) + '.bam_done'):
+            out_put_threads = pysam.Samfile(pOutputFileBufferDir + str(counter) + '.bam', 'rb')
+            while True:
+                try:
+                    data = out_put_threads.next()
+                except StopIteration:
+                    break
+                out_bam.write(data)
+            out_put_threads.close()
+            os.remove(pOutputFileBufferDir + str(counter) + '.bam')
+            os.remove(pOutputFileBufferDir + str(counter) + '.bam_done')
+            counter += 1
+        else:
+            time.sleep(3)
 
-        # prepare data for bam output
-        # set the flag to point that this data is paired
-        mate1.flag |= 0x1
-        mate2.flag |= 0x1
-
-        # set one read as the first in pair and the
-        # other as second
-        mate1.flag |= 0x40
-        mate2.flag |= 0x80
-
-        # set chrom of mate
-        mate1.mrnm = mate2.rname
-        mate2.mrnm = mate1.rname
-
-        # set position of mate
-        mate1.mpos = mate2.pos
-        mate2.mpos = mate1.pos
-
-        out_bam.write(mate1)
-        out_bam.write(mate2)
-
-        if iter_num % 5e6 == 0:
-            # every 5 million iterations append to the matrix
-            # otherwise the row, col and data vectors continue growing and
-            # for a large dataset the system could run out of memory
-            if hic_matrix is None:
-                hic_matrix = coo_matrix((data, (row, col)), shape=(matrix_size, matrix_size))
-            else:
-                hic_matrix += coo_matrix((data, (row, col)), shape=(matrix_size, matrix_size))
-            row = []
-            col = []
-            data = []
+    out_bam.close()
+    os.remove(pOutputFileBufferDir + 'done_processing')
 
-    if hic_matrix is None:
-        hic_matrix = coo_matrix((data, (row, col)), shape=(matrix_size, matrix_size))
+
+def main(args=None):
+    """
+    Reads line by line two bam files that are not sorted.
+    Each line in the two bam files should correspond
+    to the mapped position of the two ends of a Hi-C
+    fragment.
+
+    Each mate pair is assessed to determine if it is
+    a valid Hi-C pair, in such case a matrix
+    reporting the counts of mates is constructed.
+
+    A bam file containing the valid Hi-C reads
+    is also constructed
+    """
+
+    # check pysam version
+    if LooseVersion(pysam.__version__) < LooseVersion("0.8.3"):
```

Comment:
I would remove this from the code, the requirement.txt file should take care of this.